### PR TITLE
Token-aware text wrapping of commit summaries

### LINKED ISSUE #9185: urls within commits get cut off and cannot get opened by the browser
### Describe the bug

lookup a commits within the history of a project which contains a url.
the url - when displayed cut off - because of its length does not open properly in the browser

### Version & OS

Windows10x64
GHD Version 2.3.1

### Steps to reproduce the behavior

1. clone https://github.com/phpstan/phpstan-src
2. open the history
3. select the commit 7d2ecf9a1f911f4bc3d64373119097e927084221
4. click the url contained in the commit

image

### Expected behavior

the url opens in the browser

### Actual behavior

the browser does not open the actual url from the commit `https://github.com/phpstan/phpstan/issues/3008` but it seem to open the cut-off version of it `https://github.com/phpstan/phpstan/i%E2%80%A6` -> which leads to a 404 http error page on github.com

opening the same url from within github.com works like expected, see https://github.com/phpstan/phpstan-src/commit/7d2ecf9a1f911f4bc3d64373119097e927084221

### PR DESCRIPTION
**Pull request by niik**
Closes #9185 

## Description

As described in https://github.com/desktop/desktop/issues/9185 our naive text wrapping efforts implemented in https://github.com/desktop/desktop/pull/2575 suffer from a major problem when commit summaries contains links that gets wrapped.

image

Since the wrapper operates on the raw text of the commit summary before the `Tokenizer` class (used by the `RichText` component) has had a chance to work it happily hard wraps in the middle of a url. When the `RichText` component then gets a chance to work on it it creates a link out of the link that remains.

This was a limitation that was called out in #2575

> [...] a fairly simple fix we think will work alright for now. Note that this might cause links to break if we truncate them in the middle. The solution for this would be to do a pre-pass with the tokenizer and truncate using tokens instead but that's a bigger task that I'm not convinced that we need right now.

In reviewing #9259 it became clear that that assumption no longer held true and we needed to look into a robust way of doing wrapping that was Token aware.

In this PR I've made it possible to provide the `RichText` component with a list of tokens as well as a string for the component to tokenize. I've also updated the `CommitSummary` to tokenize both the commit summary and the body before providing them to the `RichText` component.

Finally I've implemented a token-aware wrapper function that hopefully covers most of the many edge-cases involved in doing this type of wrapping. I've set up https://github.com/niik/commit-summary-wrap-tests/commits/master as a list of commits with weird behaviors that should be accounted for and intend to keep adding commits there as I encounter them.

### Screenshots

Here's a few before and afters

**Before**
image

**After**

image

Here the wrapper takes into account the fact that the link has been resolved as an issue link and thereby shortened to just two characters.

**Before**

image

**After**

image

**Before**

image

This  illustrates the main problem described in #9185.

**After**

image

While the link is hard-wrapped both parts are clickable and go to the same URL.

**Before**

image

Not aware of emojis, splits after 72 characters (counting ':one:` as five characters)

**After**

image

Emojis are counted as two regular characters so we're  able to fit a lot of them in  the commit summary.

## Notes

This does not solve (but it doesn't make the situation any worse either) the problem of multibyte unicode characters (native emoji). They can still get wrapped in the middle. I don't intend to address that as part of this PR.

## Release notes

Notes: [Fixed] Links are no longer broken when text wrapping needs to take place in commit summaries

### CODE REVIEW

File: app/src/lib/text-token-parser.ts [hunk 1]
```diff
@@ -338,7 +336,7 @@ export class Tokenizer {
    *
    * @returns an array of tokens representing the scan results.
    */
-  public tokenize(text: string): ReadonlyArray<TokenResult> {
+  public tokenize(text: string): TokenResult[] {
```
Comment (rafeca) [inline app/src/lib/text-token-parser.ts:339]: Since this is a public method and it's used from React components I'm not super happy about making the output array mutable (with these current changes it's still safe but I'm worried that at some point somebody will pass this mutable around as a Prop to a React component).

Can we keep this as a readonly array? (see my comment below)

File: app/src/lib/wrap-rich-text-commit-message.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      body.unshift({ kind: TokenType.Text, text: `…` }, ...overflow, {
+        kind: TokenType.Text,
+        text: '\n\n',
+      })
+    } else {
+      body.unshift({ kind: TokenType.Text, text: `…` }, ...overflow)
+    }
+  }
+
+  return { summary, body }
```
Comment (rafeca) [inline app/src/lib/wrap-rich-text-commit-message.ts:116]: Maybe we can avoid mutating the output of the tokenizer by using early returns:

```suggestion
  if (overflow.length === 0) {
    return { summary, body }
  }

  summary.push({ kind: TokenType.Text, text: '…' })
  overflow.unshift({ kind: TokenType.Text, text: `…` })

  if (body.length === 0) {
    return { summary, body: overflow }
  }

  return {
    summary,
    body: [...overflow, { kind: TokenType.Text, text: `…` }, ...body],
  }
```

(another option would be to just clone the output array and keep the current logic).

File: app/src/lib/text-token-parser.ts (hunk 1 above)
Comment (niik) [reply app/src/lib/text-token-parser.ts]: > at some point somebody will pass this mutable around as a Prop to a React component

You sure know how to nerd-snipe me. See https://github.com/desktop/desktop/pull/9939

> Can we keep this as a readonly array? (see my comment below)

Yep, I took a stab at reworking it, I took a slightly different approach to yours but I think we achieve the same end-result.

File: app/src/lib/wrap-rich-text-commit-message.ts (hunk 2 above)
Comment (niik) [reply app/src/lib/wrap-rich-text-commit-message.ts]: Took a stab at this, feel free to have another look

**Comment by niik** [comment]
Thanks y'all!
